### PR TITLE
 Common/Util: Introduce a task scheduler which schedules task with c++11 and 17 techniques

### PR DESCRIPTION
**Pull request by Viste**
#### CHANGES PROPOSED:
-  Use std::function as callback (methods, lamdas & std::bind).
-  Uses std::chrono::duration as duration type (Milliseconds, Seconds...) with c++11 user defined literals we could just write "11s" for example (MSVS 2015, gcc & clang).
- Provides a lot of overloads for easier usage:
  - Static durations or random durations between min and max.
  - Named (with group ids - group id 0 is possible) or anonymous tasks.
  - Asynchronous tasks.
- Is adapted to the current needs of scripting:
  - Repeat tasks through the TaskContext
  - Schedule new tasks from the context.
  - Provides access to the times repeat counter of a task,
    this will help to schedule multi step events (dialogs for example).

See https://gist.github.com/Naios/6d143538abdc3d32fd2b for usage examples.

Big thanks to @Naios

also move scheduled map scripts counter methods into MapManage

##### Target branch(es): master

### CODE REVIEW

**Comment by FrancescoBorzi** [comment]
@Viste the build is broken, also about:

> dont know how you can test it

just mention a list of things that this can potentially affect, so people can check those

**Comment by Viste** [comment]
@FrancescoBorzi need C++17 to fix build ))) no <Optional> in C++11/14 std

**Comment by FrancescoBorzi** [comment]
waiting for @Winfidonarleyan to fix C++17 in Windows : https://github.com/azerothcore/azerothcore-wotlk/pull/2234

I don't have Windows and can't do anything

**Comment by Viste** [comment]
@FrancescoBorzi same for me xD i even run battlenet and WoW classic through wine

**Comment by Viste** [comment]
i think need change auto_ptr -> unique_ptr

**Comment by Viste** [comment]
@FrancescoBorzi ready for test

**Comment by FrancescoBorzi** [comment]
@Viste 

> HOW TO TEST THE CHANGES: dont know how you can test it

can you mention some things that might be affected so they can be checked by testers?

**Comment by pklloveyou** [comment]
TIM截图20190903212421

**Comment by pklloveyou** [comment]
TIM截图20190903212421

**Comment by Viste** [comment]
???

**Comment by pklloveyou** [comment]
E0312	There is no user defined "std::_Mem_fn<TaskScheduler &(TaskScheduler::*)()>" to "const std::function<TaskScheduler &(TaskScheduler &)>" Proper conversion of

**Comment by Viste** [comment]
okay its a VS 2015 bug >_<

**Comment by pklloveyou** [comment]
The problem is the inability to convert " const"

**Comment by pklloveyou** [comment]
How to test it?

**Comment by Viste** [comment]
@pklloveyou first you need test some instances all should work as indeed   
then try to write new or rewrite existed script by this example:  https://gist.github.com/Naios/6d143538abdc3d32fd2b

**Comment by pklloveyou** [comment]
I can't open this page.

**Comment by pklloveyou** [comment]
I'm testing all the ICC bosses right now.

**Comment by Viste** [comment]
hmm maybe this https://gist.github.com/Viste/41a8cd479396add7b955482ed9c5ece6

**Comment by pklloveyou** [comment]
China blocked a lot of foreign websites, also can not open

**Comment by Viste** [comment]
@pklloveyou okey give me a link to paste or something what china gov dont block

**Comment by pklloveyou** [comment]
Can't understand what you're talking about.。。

**Comment by pklloveyou** [comment]
I'm testing all the ICC bosses.

**Comment by Viste** [comment]
https://github.com/Viste/test/blob/master/README.md

**Comment by pklloveyou** [comment]
ok

**Comment by Viste** [comment]
>_< good please test it someone

**Comment by Winfidonarleyan** [comment]
After 1-3 days

**Comment by BarbzYHOOL** [comment]
we need the doc associated with this too else it will be forgotten

**Comment by Viste** [comment]
@BarbzYHOOL plz read whole topic i give link to doc

**Comment by Winfidonarleyan** [comment]
Work witch https://github.com/azerothcore/azerothcore-wotlk/pull/2240
image

Use example https://gist.github.com/Winfidonarleyan/f774b571d2f81073a6f02ef2ec9402c8

**Comment by FrancescoBorzi** [comment]
@Winfidonarleyan you tested this?

**Comment by Winfidonarleyan** [comment]
Yes, @FrancescoBorzi

File: src/common/Utilities/TaskScheduler.cpp [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+
+TaskContext& TaskContext::SetGroup(TaskScheduler::group_t const group)
+{
+    _task->_group = group;
+    return *this;
+}
+
+TaskContext& TaskContext::ClearGroup()
+{
+    _task->_group = std::nullopt;
```
Comment (LichKing255) [inline src/common/Utilities/TaskScheduler.cpp]: std has no member nullopt

File: src/common/Utilities/TaskScheduler.h [hunk 2]
```diff
@@ -0,0 +1,639 @@
...
+    typedef std::function<void()> success_t;
+
+    class Task
+    {
+        friend class TaskContext;
+        friend class TaskScheduler;
+
+        timepoint_t _end;
+        duration_t _duration;
+        std::optional<group_t> _group;
```
Comment (LichKing255) [inline src/common/Utilities/TaskScheduler.h]: what is this optional ? not found. need Boost library ?

Comment (Winfidonarleyan) [reply src/common/Utilities/TaskScheduler.h]: boot no need for this

Comment (Stoabrogga) [reply src/common/Utilities/TaskScheduler.h]: https://en.cppreference.com/w/cpp/utility/optional

File: src/common/Utilities/TaskScheduler.cpp (hunk 1 above)
Comment (Stoabrogga) [reply src/common/Utilities/TaskScheduler.cpp]: https://en.cppreference.com/w/cpp/utility/optional/nullopt

Comment (Viste) [reply src/common/Utilities/TaskScheduler.cpp]: update yr data

File: src/common/Utilities/TaskScheduler.h (hunk 2 above)
Comment (Viste) [reply src/common/Utilities/TaskScheduler.h]: update yr data base C++17 released

**Pull request closed without merging**